exporter/datadogexporter/internal/logs/translator.go

```diff
@@ -0,0 +1,172 @@
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	conventions "go.opentelemetry.io/collector/semconv/v1.6.1"
+)
+
+const (
+	namespace          = "otel"
+	otelTraceID        = namespace + ".trace_id"
+	otelSpandID        = namespace + ".span_id"
+	otelSeverityNumber = namespace + ".serverity_number"
+	otelSeverityText   = namespace + ".serverity_text"
+	otelTimestamp      = namespace + ".timestamp"
+
+	ddStatus    = "status"
+	ddTraceID   = "dd.trace_id"
+	ddSpanID    = "dd.span_id"
+	ddTimestamp = "@timestamp"
+)
+
+const (
+	logLevelTrace = "trace"
+	logLevelDebug = "debug"
+	logLevelInfo  = "info"
+	logLevelWarn  = "warn"
+	logLevelError = "error"
+	logLevelFatal = "fatal"
+)
+
+// Transform is responsible to convert LogRecord to datadog format
+func Transform(lr plog.LogRecord, res pcommon.Resource) datadogV2.HTTPLogItem {
+	hostName, serviceName := extractHostNameService(res.Attributes(), lr.Attributes())
+
+	l := datadogV2.HTTPLogItem{
+		AdditionalProperties: make(map[string]string),
+	}
+	if hostName != "" {
+		l.Hostname = datadog.PtrString(hostName)
+	}
+
+	if serviceName != "" {
+		l.Service = datadog.PtrString(serviceName)
+	}
+
+	// we need to set log attributes as AdditionalProperties
+	lr.Attributes().Range(func(k string, v pcommon.Value) bool {
+		l.AdditionalProperties[k] = v.AsString()
+		return true
+	})
+	if !lr.TraceID().IsEmpty() {
+		l.AdditionalProperties[ddTraceID] = convertTraceID(lr.TraceID().HexString())
+		l.AdditionalProperties[otelTraceID] = lr.TraceID().HexString()
+	}
+	if !lr.SpanID().IsEmpty() {
+		l.AdditionalProperties[ddSpanID] = convertTraceID(lr.SpanID().HexString())
+		l.AdditionalProperties[otelSpandID] = lr.SpanID().HexString()
+	}
+	var status string
+
+	if lr.SeverityText() != "" {
+		status = lr.SeverityText()
+		l.AdditionalProperties[otelSeverityText] = lr.SeverityText()
+	} else {
+		status = deriveStatus(lr.SeverityNumber())
+	}
+	l.AdditionalProperties[ddStatus] = status
+
+	if lr.SeverityNumber() != 0 {
+		l.AdditionalProperties[otelSeverityNumber] = fmt.Sprintf("%d", lr.SeverityNumber())
+	}
+
+	// for datadog to use the same timestamp we need to set the additional property of "@timestamp"
+	if lr.Timestamp() != 0 {
+		// we are retaining the nano second precision in this property
+		l.AdditionalProperties[otelTimestamp] = fmt.Sprintf("%d", lr.Timestamp())
+		l.AdditionalProperties[ddTimestamp] = lr.Timestamp().AsTime().Format(time.RFC3339)
+	}
+
+	var tags = attributes.TagsFromAttributes(res.Attributes())
+	if len(tags) > 0 {
+		tagStr := strings.Join(tags, ",")
+		l.Ddtags = datadog.PtrString(tagStr)
+	}
+	return l
+}
+
+func extractHostNameService(resourceAttrs pcommon.Map, logAttrs pcommon.Map) (hostName string, serviceName string) {
+	if src, ok := attributes.SourceFromAttributes(resourceAttrs, true); ok && src.Kind == source.HostnameKind {
+		hostName = src.Identifier
+	}
+
+	// hostName, serviceName are blank from resource
```

Comment:
It looks like at this point in the function we have only checked the resource for the hostName -- the serviceName part comes below. I would look at rearranging this.